sdks/java/core/src/main/java/org/apache/beam/sdk/io/UnboundedSource.java

```diff
@@ -93,6 +93,13 @@ public boolean requiresDeduping() {
     return false;
   }
 
+  /**
+   * Returns whether this source is configured for offset-based deduplication by the runner.
+   */
+  public boolean offsetDeduplication() {
```

Comment:
can this be named as a property of the source instead? The difference being that it is available to runners to use for optimization if desired but it isn't required that the runner does.

so maybe this should be named isOffsetBased with a comment that if it returns true then the source needs to provide offsets that are unique for each element, and can be ordered lexicographically for both the UnboundedReaders it creates and the checkpoint marks it vends must have an offset greater orr equal to all elements read and less than the next element (or whatever the edge-case semantics are).

Reply:
Updated to isOffsetDeduplication with more detailed comment.